### PR TITLE
[OpenReg][Feat] Add OpenReg profiler integration tests and implementation

### PR DESCRIPTION
**Pull request by KarhouTam**
## Summary
This PR adds a minimal example integration of a PRIVATEUSE1 backend ("OpenReg") to the PyTorch profiler via the legacy autograd path. OpenReg simulates a device and uses CPU timing to demonstrate how third-party/private accelerators can hook into the profiler without native Kineto support.

This update clarifies preserved, potentially unstable changes and enumerates validation scope, limitations, and follow-ups.

## What’s included

- C++ profiler stubs and observer for PRIVATEUSE1 (OpenReg)
  - Implements `ProfilerStubs` for OpenReg to record events, compute elapsed times, synchronize, and iterate devices.
  - Records timestamps using c10’s approximate CPU clock when requested; device events use OpenReg stream/event wrappers.
  - Hooks into `at::RecordFunction` scopes so aten ops and `record_function` ranges are profiled.
  - Files:
    - `test/cpp_extensions/open_registration_extension/torch_openreg/csrc/profiler/stubs/openreg.cpp`

- Integration tests exercising legacy autograd profiler with OpenReg
  - Coverage: basic ops, nested `record_function`, multiple streams/devices, synchronization, shape capture, table output, chrome trace export, backward, training/inference, performance sanity.
  - File:
    - `test/cpp_extensions/open_registration_extension/torch_openreg/tests/test_profiler.py`

## Rationale
- PRIVATEUSE1 is the intended integration point for custom/private accelerators.
- Providing a working template reduces friction for third-party backends to integrate with profiler lifecycle and RecordFunction callbacks without requiring Kineto.

## How to use (example)
```python
from torch.autograd.profiler import profile as autograd_profile

with autograd_profile(use_device="openreg", use_kineto=False, record_shapes=True) as prof:
    x = torch.randn(10, 10, device="openreg")
    y = torch.randn(10, 10, device="openreg")
    z = x @ y

print(prof.key_averages().table(sort_by="cpu_time_total", row_limit=10))
prof.export_chrome_trace("trace.json")
```

## Limitations
- No Kineto timeline or device-native markers for OpenReg.
- Timing is simulated; suitable for template/example but not for hardware benchmarking.
- Annotation APIs are no-ops (no NVTX-equivalent in this path).

## Backward compatibility
- Changes are confined to the test cpp extension and example backend; core profiler behavior for existing backends is unaffected.

## Follow-ups
- Documentation for OpenReg integration points and a concise migration guide for real PRIVATEUSE1 backends.
- Integrate OpenReg profiler via Kineto after Kineto dynamic plugin feature merged and stable.
- Additional targeted tests: phase isolation under scheduler control, more stress on stream correlation.

### CODE REVIEW

File: test/cpp_extensions/open_registration_extension/torch_openreg/CMakeLists.txt [hunk 1]
```diff
@@ -30,6 +30,8 @@ set(CMAKE_INSTALL_MESSAGE NEVER)
 set(Torch_DIR ${PYTORCH_INSTALL_DIR}/share/cmake/Torch)
 find_package(Torch REQUIRED)
 
+include_directories(${PYTORCH_INSTALL_DIR}/../third_party/kineto/libkineto/include)
```
Comment (fffrog) [inline test/cpp_extensions/open_registration_extension/torch_openreg/CMakeLists.txt]: Hi, could you please add a comment to mark this line of code as something that should be removed later? Because in my opinion, PyTorch should provide header files for Kineto, even though it doesn't currently, because Kineto is indeed very complex, and the CPU's default PyTorch compilation doesn't use the USE_KINETO flag.

**Review by fffrog — APPROVED** [review]
SGTM, thank you.

We need to monitor the progress of the refactoring to support the new integration method via dynamic libraries. Once completed, we'll try switching to the Kineto solution.

**Comment by KarhouTam** [comment]
Hi, @fffrog. I found that there are some issues with the `libkineto` header indexing in the `torch_openreg` building. To avoid making things too complex, I preserved those `libkineto` changes, and this PR will be responsible for the initial OpenReg profiler integration instead (excluding the OpenReg self-defined callbacks).

**Comment by KarhouTam** [comment]
Hello, @ai-easy-cpu. OpenReg is presented as an example for out-of-tree accelerator integration. Similarly, this profiler serves as an example for profiler integration (it is not compatible with Kineto, unlike XPU, etc.). This profiler example follows the legacy approach of Torch profiler integration, whereas Kineto is the recommended path.

Regarding your question:

> What is the relationship between Kineto and the profiler?

There is no relationship between the two. As mentioned earlier, this profiler follows the legacy approach, meaning it collects profiling data through device-defined events (you can refer to the code for further details). The recommended approach is to integrate your profiler into Kineto using a plugin method, as demonstrated by XPU and AIU.

> Will it replace the profiler, or is it just a GUI plugin or something?

Not at all. This profiler is quite simple and is intended to serve as an example to provide insights and ideas for profiler developers who wish to integrate their profiler into PyTorch.

Please feel free to ask further questions!
